In [1]:
import warnings
warnings.filterwarnings('ignore')

In [2]:
import pandas as pd

df = pd.read_csv("Women_s_E-Commerce_Clothing_Reviews_1594_1.csv", sep=";")

df.head()

Unnamed: 0  Clothing.ID  Age                    Title  \
0           1          767   33                      NaN   
1           2         1080   34                      NaN   
2           3         1077   60  Some major design flaws   
3           4         1049   50         My favorite buy!   
4           5          847   47         Flattering shirt   

                                         Review.Text  Rating  Recommended.IND  \
0  Absolutely wonderful - silky and sexy and comf...       4                1   
1  Love this dress!  it's sooo pretty.  i happene...       5                1   
2  I had such high hopes for this dress and reall...       3                0   
3  I love, love, love this jumpsuit. it's fun, fl...       5                1   
4  This shirt is very flattering to all due to th...       5                1   

   Positive.Feedback.Count   Division.Name Department.Name Class.Name  
0                        0       Initmates        Intimate  Intimates  
1                        4         General         Dresses    Dresses  
2                        0         General         Dresses    Dresses  
3                        0  General Petite         Bottoms      Pants  
4                        6         General            Tops    Blouses

In [3]:
df.drop("Unnamed: 0", axis=1, inplace=True)
df.head()

Clothing.ID  Age                    Title  \
0          767   33                      NaN   
1         1080   34                      NaN   
2         1077   60  Some major design flaws   
3         1049   50         My favorite buy!   
4          847   47         Flattering shirt   

                                         Review.Text  Rating  Recommended.IND  \
0  Absolutely wonderful - silky and sexy and comf...       4                1   
1  Love this dress!  it's sooo pretty.  i happene...       5                1   
2  I had such high hopes for this dress and reall...       3                0   
3  I love, love, love this jumpsuit. it's fun, fl...       5                1   
4  This shirt is very flattering to all due to th...       5                1   

   Positive.Feedback.Count   Division.Name Department.Name Class.Name  
0                        0       Initmates        Intimate  Intimates  
1                        4         General         Dresses    Dresses  
2                        0         General         Dresses    Dresses  
3                        0  General Petite         Bottoms      Pants  
4                        6         General            Tops    Blouses

In [4]:
df.columns = df.columns.str.lower()
df.columns = df.columns.str.replace('.', '_')
df.head()

clothing_id  age                    title  \
0          767   33                      NaN   
1         1080   34                      NaN   
2         1077   60  Some major design flaws   
3         1049   50         My favorite buy!   
4          847   47         Flattering shirt   

                                         review_text  rating  recommended_ind  \
0  Absolutely wonderful - silky and sexy and comf...       4                1   
1  Love this dress!  it's sooo pretty.  i happene...       5                1   
2  I had such high hopes for this dress and reall...       3                0   
3  I love, love, love this jumpsuit. it's fun, fl...       5                1   
4  This shirt is very flattering to all due to th...       5                1   

   positive_feedback_count   division_name department_name class_name  
0                        0       Initmates        Intimate  Intimates  
1                        4         General         Dresses    Dresses  
2                        0         General         Dresses    Dresses  
3                        0  General Petite         Bottoms      Pants  
4                        6         General            Tops    Blouses

In [5]:
def veri_seti_ozeti(df):
    print("###########################################################################################################################")
    print(df.info())
    print("#############################################-Her bir sütundaki NaN değer sayısı-#############################################")
    print(df.isnull().sum())
    print("###########################################################################################################################")
    for column in df.columns:
        print(f"Sütun: {column}")
        print(f"Unique değerler: {df[column].unique()}")
        print(f"Unique değer sayısı: {len(df[column].unique())}")
        print("###########################################################################################################################")
veri_seti_ozeti(df)

###########################################################################################################################
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23486 entries, 0 to 23485
Data columns (total 10 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   clothing_id              23486 non-null  int64 
 1   age                      23486 non-null  int64 
 2   title                    19676 non-null  object
 3   review_text              22641 non-null  object
 4   rating                   23486 non-null  int64 
 5   recommended_ind          23486 non-null  int64 
 6   positive_feedback_count  23486 non-null  int64 
 7   division_name            23472 non-null  object
 8   department_name          23472 non-null  object
 9   class_name               23472 non-null  object
dtypes: int64(5), object(5)
memory usage: 1.8+ MB
None
#############################################-Her bir sütundaki NaN değer sa

In [6]:
import nltk
import re
import string

nltk.download('stopwords')

def clean_text(text):
    text = text.lower()  
    text = re.sub(r'\[.*?\]', '', text) 
    text = re.sub(r'https?://\S+|www\.\S+', '', text)  
    text = re.sub(r'<.*?>+', '', text) 
    text = re.sub(r'[%s]' % re.escape(string.punctuation), '', text)  
    text = re.sub(r'\n', '', text)  
    text = re.sub(r'\w*\d\w*', '', text)  
    return text


df['review_text'] = df['review_text'].astype(str).apply(clean_text)

df.head()

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\is\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


clothing_id  age                    title  \
0          767   33                      NaN   
1         1080   34                      NaN   
2         1077   60  Some major design flaws   
3         1049   50         My favorite buy!   
4          847   47         Flattering shirt   

                                         review_text  rating  recommended_ind  \
0  absolutely wonderful  silky and sexy and comfo...       4                1   
1  love this dress  its sooo pretty  i happened t...       5                1   
2  i had such high hopes for this dress and reall...       3                0   
3  i love love love this jumpsuit its fun flirty ...       5                1   
4  this shirt is very flattering to all due to th...       5                1   

   positive_feedback_count   division_name department_name class_name  
0                        0       Initmates        Intimate  Intimates  
1                        4         General         Dresses    Dresses  
2                        0         General         Dresses    Dresses  
3                        0  General Petite         Bottoms      Pants  
4                        6         General            Tops    Blouses

In [7]:
from nltk.corpus import stopwords

sw = stopwords.words('english')
df['review_text'] = df['review_text'].apply(lambda x: " ".join(x for x in str(x).split() if x not in sw))

In [8]:
all_words = ' '.join(df['review_text']).split()
word_counts = pd.Series(all_words).value_counts()

filtered_word_counts = word_counts[word_counts <= 1]

df['review_text'] = df['review_text'].apply(lambda x: " ".join(x for x in x.split() if x not in filtered_word_counts))

In [9]:
import nltk
from textblob import Word

nltk.download('wordnet')
df['review_text'] = df['review_text'].apply(lambda x: " ".join([Word(word).lemmatize() for word in x.split()]))


[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\is\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


In [10]:
from nltk.sentiment import SentimentIntensityAnalyzer

sia = SentimentIntensityAnalyzer()


df['negative'] = df['review_text'].apply(lambda txt: sia.polarity_scores(txt)).apply(lambda txt: txt['neg'])
df['neutral'] = df['review_text'].apply(lambda txt: sia.polarity_scores(txt)).apply(lambda txt: txt['neu'])
df['positive'] = df['review_text'].apply(lambda txt: sia.polarity_scores(txt)).apply(lambda txt: txt['pos'])
df['compound'] = df['review_text'].apply(lambda txt: sia.polarity_scores(txt)).apply(lambda txt: txt['compound'])
df["sentiment_label"] = df["review_text"].apply(lambda x: "pos" if sia.polarity_scores(x)["compound"] >= 0.5 else "neg")

df.head()


clothing_id  age                    title  \
0          767   33                      NaN   
1         1080   34                      NaN   
2         1077   60  Some major design flaws   
3         1049   50         My favorite buy!   
4          847   47         Flattering shirt   

                                         review_text  rating  recommended_ind  \
0        absolutely wonderful silky sexy comfortable       4                1   
1  love dress sooo pretty happened find store im ...       5                1   
2  high hope dress really wanted work initially o...       3                0   
3  love love love jumpsuit fun flirty fabulous ev...       5                1   
4  shirt flattering due adjustable front tie perf...       5                1   

   positive_feedback_count   division_name department_name class_name  \
0                        0       Initmates        Intimate  Intimates   
1                        4         General         Dresses    Dresses   
2                        0         General         Dresses    Dresses   
3                        0  General Petite         Bottoms      Pants   
4                        6         General            Tops    Blouses   

   negative  neutral  positive  compound sentiment_label  
0     0.000    0.154     0.846    0.8991             pos  
1     0.000    0.503     0.497    0.9710             pos  
2     0.038    0.692     0.270    0.9081             pos  
3     0.179    0.183     0.638    0.9437             pos  
4     0.000    0.494     0.506    0.9062             pos

In [11]:
def polarity_score(compound):
    if compound > 0.05:
        return "positive"
    elif compound < -0.5:
        return "negative"
    elif compound >= -0.05 and compound < 0.05:
        return "neutral"
    

df['sentiment'] = df['compound'].apply(lambda val: polarity_score(val))
df.head()

clothing_id  age                    title  \
0          767   33                      NaN   
1         1080   34                      NaN   
2         1077   60  Some major design flaws   
3         1049   50         My favorite buy!   
4          847   47         Flattering shirt   

                                         review_text  rating  recommended_ind  \
0        absolutely wonderful silky sexy comfortable       4                1   
1  love dress sooo pretty happened find store im ...       5                1   
2  high hope dress really wanted work initially o...       3                0   
3  love love love jumpsuit fun flirty fabulous ev...       5                1   
4  shirt flattering due adjustable front tie perf...       5                1   

   positive_feedback_count   division_name department_name class_name  \
0                        0       Initmates        Intimate  Intimates   
1                        4         General         Dresses    Dresses   
2                        0         General         Dresses    Dresses   
3                        0  General Petite         Bottoms      Pants   
4                        6         General            Tops    Blouses   

   negative  neutral  positive  compound sentiment_label sentiment  
0     0.000    0.154     0.846    0.8991             pos  positive  
1     0.000    0.503     0.497    0.9710             pos  positive  
2     0.038    0.692     0.270    0.9081             pos  positive  
3     0.179    0.183     0.638    0.9437             pos  positive  
4     0.000    0.494     0.506    0.9062             pos  positive

In [12]:
from sklearn.preprocessing import LabelEncoder

df["sentiment_label"] = LabelEncoder().fit_transform(df["sentiment_label"])


In [13]:
X = df["review_text"]
y = df["sentiment_label"]

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)


from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer = CountVectorizer()
X_train_count = vectorizer.fit_transform(X_train)
X_test_count = vectorizer.transform(X_test)


tf_idf_word_vectorizer = TfidfVectorizer()
X_train_tf_idf_word = tf_idf_word_vectorizer.fit_transform(X_train)
X_test_tf_idf_word = tf_idf_word_vectorizer.transform(X_test)

tf_idf_ngram_vectorizer = TfidfVectorizer(ngram_range=(2, 3))
X_train_tf_idf_ngram = tf_idf_ngram_vectorizer.fit_transform(X_train)
X_test_tf_idf_ngram = tf_idf_ngram_vectorizer.transform(X_test)

In [14]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report
from sklearn.model_selection import cross_val_score

# Count Vectors
print("Count Vectors")
log_model = LogisticRegression().fit(X_train_count, y_train)
y_pred_count = log_model.predict(X_test_count)
print(classification_report(y_pred_count, y_test))
print(cross_val_score(LogisticRegression(), CountVectorizer().fit_transform(X), y, scoring="accuracy", cv=3).mean())

# TF-IDF Word-Level
print("TF-IDF Word-Level")
log_model = LogisticRegression().fit(X_train_tf_idf_word, y_train)
y_pred_tf_idf_word = log_model.predict(X_test_tf_idf_word)
print(classification_report(y_pred_tf_idf_word, y_test))
print(cross_val_score(LogisticRegression(), TfidfVectorizer().fit_transform(X), y, scoring="accuracy", cv=3).mean())

# TF-IDF N-GRAM
print("TF-IDF N-GRAM")
log_model = LogisticRegression().fit(X_train_tf_idf_ngram, y_train)
y_pred_tf_idf_ngram = log_model.predict(X_test_tf_idf_ngram)
print(classification_report(y_pred_tf_idf_ngram, y_test))
print(cross_val_score(LogisticRegression(), TfidfVectorizer(ngram_range=(2, 3)).fit_transform(X), y, scoring="accuracy", cv=3).mean())

Count Vectors
              precision    recall  f1-score   support

           0       0.68      0.79      0.73       721
           1       0.97      0.95      0.96      5151

    accuracy                           0.93      5872
   macro avg       0.83      0.87      0.84      5872
weighted avg       0.93      0.93      0.93      5872

0.9256151692347275
TF-IDF Word-Level
              precision    recall  f1-score   support

           0       0.43      0.94      0.59       386
           1       1.00      0.91      0.95      5486

    accuracy                           0.92      5872
   macro avg       0.71      0.92      0.77      5872
weighted avg       0.96      0.92      0.93      5872

0.9102443770251013
TF-IDF N-GRAM
              precision    recall  f1-score   support

           0       0.00      0.00      0.00         0
           1       1.00      0.86      0.92      5872

    accuracy                           0.86      5872
   macro avg       0.50      0.43      0.46 

Count Vectors:

Precision, recall ve f1-score değerleri sınıf 0 için 0.68, 0.79, 0.73; sınıf 1 için ise 0.97, 0.95, 0.96 olarak yüksektir.
Bu, modelin sınıf 1'i (olumlu yorumları) iyi tahmin ettiğini ancak sınıf 0'ı (olumsuz yorumları) daha az doğru tahmin ettiğini gösterir.
Accuracy %93 olarak yüksek, yani genel olarak model iyi performans gösteriyor.
Cross-validation score (0.925) da yüksek, modelin genel geçerliliğini destekliyor.

TF-IDF Word-Level:

Precision, recall ve f1-score değerleri sınıf 0 için 0.43, 0.94, 0.59; sınıf 1 için ise 1.00, 0.91, 0.95 olarak görünmektedir.
Sınıf 0 için precision düşüktür, yani sınıf 0 olarak tahmin edilenlerin çoğu gerçekten sınıf 0 değildir.
Recall sınıf 1 için düşüktür, yani gerçek sınıf 1'lerin çoğu tespit edilmemiştir.
Accuracy %92 olarak yüksek, ancak Count Vectors'a göre daha düşüktür.
Cross-validation score (0.910) da yine Count Vectors'a göre düşüktür, bu da genel geçerliliğin Count Vectors'a kıyasla daha düşük olduğunu gösterir.

TF-IDF N-GRAM:

Precision, recall ve f1-score değerleri düşüktür, sınıf 0 ve sınıf 1 için de benzer düşük değerler görülmektedir.
Model, N-GRAM özelliklerini kullanarak daha kötü performans göstermektedir.
Accuracy %86 olarak düşüktür, modelin genel olarak kötü performans gösterdiğini gösterir.
Cross-validation score (0.851) da düşüktür, bu da modelin genel geçerliliğinin zayıf olduğunu doğrular.
Sonuç olarak, Count Vectors yöntemi diğerlerine kıyasla daha iyi bir performans sergilemiştir. Daha yüksek precision, recall, f1-score ve accuracy değerleri ile daha güvenilir tahminler yapabilmektedir. Diğer yöntemlerin performansı ise belirli metin verileri setine ve sınıflandırma görevine bağlı olarak değişiklik gösterebilir.

In [15]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_score

# Count Vectors
print("Count Vectors")
rf_model = RandomForestClassifier().fit(X_train_count, y_train)
y_pred_count = rf_model.predict(X_test_count)
print(classification_report(y_pred_count, y_test))
print(cross_val_score(RandomForestClassifier(), CountVectorizer().fit_transform(X), y, scoring="accuracy", cv=3).mean())


# TF-IDF Word-Level
print("TF-IDF Word-Level")
rf_model = RandomForestClassifier().fit(X_train_tf_idf_word, y_train)
y_pred_tf_idf_word = rf_model.predict(X_test_count)
print(classification_report(y_pred_tf_idf_word, y_test))
print(cross_val_score(LogisticRegression(), TfidfVectorizer().fit_transform(X), y, scoring="accuracy", cv=3).mean())


# TF-IDF N-GRAM
print("TF-IDF N-GRAM")
rf_model = RandomForestClassifier().fit(X_train_tf_idf_ngram, y_train)
y_pred_tf_idf_ngram = rf_model.predict(X_test_tf_idf_ngram)
print(classification_report(y_pred_tf_idf_ngram, y_test))
print(cross_val_score(RandomForestClassifier(), TfidfVectorizer(ngram_range=(2, 3)).fit_transform(X), y, scoring="accuracy", cv=3).mean())


Count Vectors
              precision    recall  f1-score   support

           0       0.28      0.99      0.44       237
           1       1.00      0.89      0.94      5635

    accuracy                           0.90      5872
   macro avg       0.64      0.94      0.69      5872
weighted avg       0.97      0.90      0.92      5872

0.8923613502019044
TF-IDF Word-Level
              precision    recall  f1-score   support

           0       0.45      0.62      0.52       601
           1       0.95      0.91      0.93      5271

    accuracy                           0.88      5872
   macro avg       0.70      0.77      0.73      5872
weighted avg       0.90      0.88      0.89      5872

0.9102443770251013
TF-IDF N-GRAM
              precision    recall  f1-score   support

           0       0.46      0.53      0.50       719
           1       0.93      0.91      0.92      5153

    accuracy                           0.87      5872
   macro avg       0.70      0.72      0.71 

Count Vectors

Precision, Recall, F1-score ve Accuracy: Count vektörleri kullanılarak eğitilen model, 0 sınıfı için düşük precision (0.28), ancak yüksek recall (0.99) değerine sahiptir. Yani, 0 olarak tahmin ettiği durumların çoğu gerçekten 0'dır, ancak 0 olarak tahmin ettiği durumların yalnızca küçük bir kısmı doğrudur. Diğer yandan, 1 sınıfı için yüksek precision (1.00) ve daha düşük recall (0.89) elde edilmiştir. Bu, 1 olarak tahmin edilen durumların neredeyse tamamının doğru olduğunu, ancak 1 olan durumların küçük bir kısmının gözden kaçırıldığını gösterir.

Accuracy: 0.90, yani toplam doğru sınıflandırma oranı %90'dır.

Cross-validation Accuracy: 0.8923613502019044, yani cross-validation kullanarak elde edilen ortalama doğruluk %89.24'tür.

Count Vectors yöntemi, 1 sınıfı için yüksek precision ve düşük recall ile dikkat çekerken, 0 sınıfı için düşük precision ve yüksek recall gösterir. Bu, modelin 1 sınıfını başarıyla tanımladığını ancak 0 sınıfında bazı zorluklar yaşadığını gösterir.

TF-IDF Word-Level

Precision, Recall, F1-score ve Accuracy: TF-IDF kelime seviyesi vektörleri kullanılarak eğitilen model, 0 ve 1 sınıfları için daha dengeli sonuçlar verir. 0 sınıfı için precision (0.45) ve recall (0.62) değerleri daha düşüktür. 1 sınıfı için precision (0.95) ve recall (0.91) değerleri yüksektir.

Accuracy: 0.88, yani toplam doğru sınıflandırma oranı %88'dir.

Cross-validation Accuracy: 0.9102443770251013, yani cross-validation kullanarak elde edilen ortalama doğruluk %91.02'dir.

TF-IDF Word-Level yöntemi, genel olarak iyi bir performans gösterir ve 1 sınıfını tanımlamada daha dengeli bir yaklaşım sergiler.

TF-IDF N-GRAM

Precision, Recall, F1-score ve Accuracy: TF-IDF n-gram vektörleri kullanılarak eğitilen model, 0 ve 1 sınıfları için Count Vectors'a benzer performans sergiler. 0 sınıfı için precision (0.46) ve recall (0.53), 1 sınıfı için precision (0.93) ve recall (0.91) olarak ölçülmüştür.

Accuracy: 0.87, yani toplam doğru sınıflandırma oranı %87'dir.

Cross-validation Accuracy: 0.8642594804344848, yani cross-validation kullanarak elde edilen ortalama doğruluk %86.43'tür.

TF-IDF N-GRAM yöntemi ise 1 sınıfı için yüksek precision ve recall ile iyidir, ancak 0 sınıfı için Count Vectors ve TF-IDF Word-Level'e kıyasla biraz daha düşük performans gösterir.


Kullanılan metin özelliklerinin sınıflandırma performansına nasıl etki ettiğini anlayabilirsiniz. Count Vectors daha yüksek recall sağlarken, TF-IDF vektörleri genellikle daha yüksek precision ve F1-score değerleriyle birlikte daha dengeli sonuçlar verir.